test_plans/bfloat16.asciidoc

```diff
@@ -0,0 +1,149 @@
+:sectnums:
+:xrefstyle: short
+
+= Test plan for bfloat16
+
+This is a test plan for the APIs described in
+https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_bfloat16.asciidoc[sycl_ext_oneapi_bfloat16]
+
+== Testing scope
+
+=== Device coverage
+
+All of the tests described below are performed only on the default device that
+is selected on the CTS command line.
+
+=== Type coverage
+
+All of the tests described below are performed using each of the following `typename T`:
+
+* `char`
+* `short`
+* `int`
+* `long long`
+* `std::size_t`
+* `bool`
+* `float`
+
+In addition, if the device has aspect `aspect::fp64`:
+
+* `double`
+
+In addition, if the device has `aspect::fp16`:
+
+* `sycl::half`
+
+Note that `T` must be convertible to `float`
+
+== Tests
+
+=== Size
+
+`bfloat16` must occupy 2 bytes of memory, check that `sizeof(bfloat16) == 2`.
+This check is required to calculate dynamic range.
+
+=== Constructors
+
+`bfloat16()` +
+`bfloat16(const bfloat16 &)` +
+`~bfloat16()`
+
+Check that:
+
+* `std::is_default_constructible_v<bfloat16> == true`;
+* `std::is_copy_constructible_v<bfloat16> == true`;
+* `std::is_destructible_v<bfloat16> == true`.
+
+`bfloat16(const float &a)` +
+`bfloat16 &operator=(const float &a)`
+
+* Create a `float` variable `f` equal to 1;
+* Create a `bfloat16` variable `bf1` passing `f` to the constructor;
+* Create a `bfloat16` variable `bf2` by assinging it to `f`;
+* Verify that:
+    ** `bf1 == f`
+    ** `bf2 == f`
+
+`bfloat16(const sycl::half &a)` +
+`bfloat16 &operator=(const sycl::half &a)`
+
+Same as above, but with `sycl::half` instead of `float`
+
+=== Dynamic range
+
+* Calculate the maximum and minimum positive values of `bfloat16`:
+    ** Create `std::bitset` with 16 bits to store maximum value;
+    ** Set all bits to 1;
+    ** Set bit 15 (sign bit) and bit 7 (special values) to 0;
+    ** Convert it to `bfloat16` variable `BFLT16_MAX`
+
+    ** Create `std::bitset` with 16 bits to store minimum value;
+    ** Set all bits to 0;
+    ** Set bit 8 to 0 to get minimum normal value;
+    ** Convert it to `bfloat16` variable `BFLT16_MIN`
+
+* Calculate `float` dynamic range equal to `log10(FLT_MAX / FLT_MIN)` or `log10(FLT_MAX) - log10(FLT_MIN)`;
+* Calculate `bfloat16` dynamic range equal to `log10(BFLT16_MIN) - log10(BFLT16_MIN)`;
+* Compare integer parts of computed values.
```

Comment:
I don't understand the point of this test.  It seems like it is just testing mathematical properties of the bfloat16 floating point representation, which doesn't really test the implementation.

The only part of the implementation that you are testing is that you can convert the special values "BFLT16_MIN" and "BFLT16_MAX" to `float`.  This does seem valuable, but it is not testing the "dynamic range".

I'd suggest changing this test and calling it "Test special values".  For example, the test for BFLT16_MIN could look like:

```
uint16_t min_bits = 0b0000000010000000;
bfloat16 bfmin = sycl::bit_cast<bfloat16>(min_bits);
uint32_t float_bits = 0b00000000100000000000000000000000;
float expect = sycl::bit_cast<float>(float_bits);
if (bfmin != expect)
  /* test fails */
```

You could expand this test to other special values like positive / negative infinity and NaN.

Reply:
Do I understand correctly that dynamic range is ratio of maximum to minimum value? I took a logarithm to reduce very large values of this ratio, as audio engineers do. Spec says that dynamic range of `bfloat16` is the same as `float`, so I wanted to test it this way.

Reply:
I think this is the relevant part of the spec:

> However, bfloat16 allots 8 bits to the exponent instead of the 5 bits used by sycl::half and 7 bits to the significand versus 10 bits used by sycl::half. Thus, bfloat16 has the same dynamic range as a 32-bit float but with reduced precision.

The statement about the dynamic range is simply a result of the format, which has an 8 bit exponent and a 7 bit significand.  If you want to test this sentence in the spec, you should somehow test that the format has an 8 bit exponent and a 7 bit significand.

The test I propose above would do this because it constructs the value of a `bfloat16` number by setting specific bits in the exponent and significand and then comparing the `bfloat16` number with the expected value of a `float`.  This comparison will fail if the implementation does not use an 8 bit exponent and 7 bit significand.

Reply:
Done.